### PR TITLE
Google tasks - New Actions

### PR DESCRIPTION
**Pull request by vellames-turing**
- [x] List Tasks | Returns a list of all tasks in the specified task list
- [x] Create a Task | Creates a new task
- [x] Update a Task | Updates an existing task
- [x] Delete a Task | Deletes a task
- [x] List Task Lists | Returns a list of all task lists
- [x] Create a Task List | Creates a new task list
- [x] Update a Task List | Updates an existing task list
- [x] Delete a Task List | Deletes a task list

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/google_tasks/actions/create-task/create-task.mjs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-create-task",
+  name: "Create Task",
+  description: "Creates a new task and adds it to the authenticated user's task lists.  [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasks/insert)",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/create-task/create-task.mjs:6]: Extra space

```suggestion
  description: "Creates a new task and adds it to the authenticated user's task lists. [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasks/insert)",
```

File: components/google_tasks/google_tasks.app.mjs [hunk 2]
```diff
@@ -1,11 +1,167 @@
...
+    _getHeaders() {
+      return {
+        "Content-Type": "application/json",
+        "Authorization": `Bearer ${this.$auth.oauth_access_token}`,
+      };
+    },
+    _getRequestParams(opts = {}) {
+      return {
+        ...opts,
+        url: this._getBaseUrl() + opts.path,
```
Comment (andrewjschuang) [inline components/google_tasks/google_tasks.app.mjs:84]: If `opts.path` is undefined, `url` will break here

File: components/google_tasks/google_tasks.app.mjs [hunk 3]
```diff
@@ -1,11 +1,167 @@
...
+        const pageResult = await fn(this, {
+          ...params,
+          maxResults: ITEMS_PER_PAGE,
+          pageToken: nextPageToken,
+        }, ...rest);
+        nextPageToken = pageResult.nextPageToken;
+        data = [
+          ...data,
+          ...pageResult.items,
+        ];
```
Comment (andrewjschuang) [inline components/google_tasks/google_tasks.app.mjs:103]:
```suggestion
        data.push(...pageResult.items);
```

File: components/google_tasks/google_tasks.app.mjs [hunk 4]
```diff
@@ -1,11 +1,167 @@
...
+      return {
+        ...opts,
+        url: this._getBaseUrl() + opts.path,
+        headers: this._getHeaders(),
+      };
+    },
+    async paginate(fn, params, ...rest) {
+      let data = [];
+      let nextPageToken = null;
+      const TOTAL_MAX_RESULTS = Math.min(params.maxResults, 10000);
```
Comment (andrewjschuang) [inline components/google_tasks/google_tasks.app.mjs:91]: I think `10000` may be too much for one step. `200-1000` might be fine

File: components/google_tasks/google_tasks.app.mjs [hunk 5]
```diff
@@ -1,11 +1,167 @@
...
+            label: task.title,
+            value: task.id,
+          })),
+        };
+      },
+    },
+    completed: {
+      type: "boolean",
+      label: "Completed",
+      description: "Mark as `true` if you task is already completed.",
```
Comment (andrewjschuang) [inline components/google_tasks/google_tasks.app.mjs:62]:
```suggestion
      description: "Mark as `true` if your task is already completed.",
```

File: components/google_tasks/google_tasks.app.mjs [hunk 6]
```diff
@@ -1,11 +1,167 @@
...
+  propDefinitions: {
+    title: {
+      type: "string",
+      label: "Title",
+      description: "The title of the task list.",
+    },
+    maxResults: {
+      type: "integer",
+      label: "Max Results",
+      description: "Maximum number of register to be fetched.",
```
Comment (andrewjschuang) [inline components/google_tasks/google_tasks.app.mjs:15]:
```suggestion
      description: "Maximum number of registers to be fetched.",
```

File: components/google_tasks/actions/create-task-list/create-task-list.mjs [hunk 7]
```diff
@@ -0,0 +1,25 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-create-task0list",
+  name: "Create Task List",
+  description: "Creates a new task list and adds it to the authenticated user's task lists.  [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasklists/insert)",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/create-task-list/create-task-list.mjs:6]: Extra space

```suggestion
  description: "Creates a new task list and adds it to the authenticated user's task lists. [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasklists/insert)",
```

File: components/google_tasks/actions/create-task-list/create-task-list.mjs [hunk 8]
```diff
@@ -0,0 +1,25 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-create-task0list",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/create-task-list/create-task-list.mjs:4]:
```suggestion
  key: "google_tasks-create-task-list",
```

File: components/google_tasks/actions/delete-task-list/delete-task-list.mjs [hunk 9]
```diff
@@ -0,0 +1,23 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-delete-task-list",
+  name: "Delete Task List",
+  description: "Deletes the authenticated user's specified task list.  [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasklists/delete)",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/delete-task-list/delete-task-list.mjs:6]: Extra space

```suggestion
  description: "Deletes the authenticated user's specified task list. [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasklists/delete)",
```

File: components/google_tasks/actions/delete-task/delete-task.mjs [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-delete-task",
+  name: "Delete Task",
+  description: "Deletes the authenticated user's specified task.  [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasks/delete)",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/delete-task/delete-task.mjs:6]: Extra space

```suggestion
  description: "Deletes the authenticated user's specified task. [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasks/delete)",
```

File: components/google_tasks/actions/list-tasks/list-tasks.mjs [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    maxResults: {
+      propDefinition: [
+        app,
+        "maxResults",
+      ],
+    },
+    showCompleted: {
+      type: "boolean",
+      label: "Show Completed",
+      description: "Whether completed tasks are returned in the result. Optional. Defaults to `true`.",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/list-tasks/list-tasks.mjs:26]:
```suggestion
      description: "Whether completed tasks are returned in the result. Defaults to `true`.",
```

File: components/google_tasks/actions/list-tasks/list-tasks.mjs [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+      optional: true,
+    },
+    showDeleted: {
+      type: "boolean",
+      label: "Show Deleted",
+      description: "Whether deleted tasks are returned in the result. Optional. Defaults to `false`.",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/list-tasks/list-tasks.mjs:32]:
```suggestion
      description: "Whether deleted tasks are returned in the result. Defaults to `false`.",
```

File: components/google_tasks/actions/list-tasks/list-tasks.mjs [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+      maxResults: this.maxResults,
+      showCompleted: this.showCompleted,
+      showDeleted: this.showDeleted,
+    };
+    const res = await this.app.paginate(
+      this.app.getTasks.bind(this),
+      params,
+      this.taskListId,
+    );
+    $.export("$summary", "Action successfully performed");
```
Comment (andrewjschuang) [inline components/google_tasks/actions/list-tasks/list-tasks.mjs:47]: A bit more clear

```suggestion
    $.export("$summary", "Listed tasks successfully");
```

File: components/google_tasks/actions/list-tasks-list/list-tasks-list.mjs [hunk 14]
```diff
@@ -0,0 +1,29 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-list-tasks-list",
+  name: "List Tasks List",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/list-tasks-list/list-tasks-list.mjs:5]: Also change `filename` and `key`

```suggestion
  name: "List Task Lists",
```

File: components/google_tasks/actions/list-tasks-list/list-tasks-list.mjs [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+  },
+  async run({ $ }) {
+    const params = {
+      maxResults: this.maxResults,
+    };
+    const res = await this.app.paginate(
+      this.app.getTaskLists.bind(this),
+      params,
+    );
+    $.export("$summary", "List(s) Task successfully retrieved");
```
Comment (andrewjschuang) [inline components/google_tasks/actions/list-tasks-list/list-tasks-list.mjs:26]:
```suggestion
    $.export("$summary", "Task List(s) successfully retrieved");
```

File: components/google_tasks/actions/update-task/update-task.mjs [hunk 16]
```diff
@@ -0,0 +1,65 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-update-task",
+  name: "Update Task",
+  description: "Updates the authenticated user's specified task.  [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasks/update)",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/update-task/update-task.mjs:6]: Extra space

```suggestion
  description: "Updates the authenticated user's specified task. [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasks/update)",
```

File: components/google_tasks/actions/update-task-list/update-task-list.mjs [hunk 17]
```diff
@@ -0,0 +1,32 @@
+import app from "../../google_tasks.app.mjs";
+
+export default {
+  key: "google_tasks-update-task-list",
+  name: "Update Task List",
+  description: "Updates the authenticated user's specified task list.  [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasklists/update)",
```
Comment (andrewjschuang) [inline components/google_tasks/actions/update-task-list/update-task-list.mjs:6]:
```suggestion
  description: "Updates the authenticated user's specified task list. [See the docs here](https://developers.google.com/tasks/reference/rest/v1/tasklists/update)",
```

**Comment by vellames-turing** [comment]
Hi there, tks for reviewing it! All suggestions addressed

**Review by andrewjschuang — APPROVED** [review]

**Comment by andrewjschuang** [comment]
Great, thanks for the changes!

**Comment by DilanAthukorala** [comment]
Verification completed and no issues found.

**Pull request merged**